Adaptive Routing and Switching

JaiOCP opened the pull request.

mickeyspiegel commented:
The diffs for inc/saitam.h are a bit messy, not aligned to where the changes were actually made. It would be easier to just view the file and search for TAM_INT.

mickeyspiegel commented:
Rebased the proposal on latest JaiOCP:patch-16 that removed the other SAI TAM INT proposal since it belongs in a separate pull request. The rebase clarifies what additions are required on top of the agreed SAI TAM content in JaiOCP:patch-16, in order to add SAI TAM INT functionality.

In addition, the commit message was reworked to point out differences between the two SAI TAM INT proposals:

1. Flow selection

   While both proposals now agree on the use of ACLs to select the subset
   of traffic to which INT applies, there are significant differences in
   how the ACLs and INT parameters are bound together.

   In this proposal, the ACL refers to a "TAM INT Group" object that
   specifies relevant attributes. By simply following the reference, one
   can easily determine all the details of the INT insertion action that
   is applied to traffic matching this ACL entry. Different entries in
   the same ACL may refer to different "TAM INT Group" objects, if
   different values of any attributes are desired such as trace type /
   instruction bitmap, max hop length, namespace ID, etc.

   The ACL may have binding points to ports, VLANs, etc using the
   existing references from those entities to ACLs or ACL groups.
   Resolution between overlapping ACL entries within the same ACL is
   done through comparison of the priority of the ACL entries, as is
   normal for any type of ACL.

   The other SAI TAM INT proposal creates a "TAM INT" object that, among
   other things, specifies attributes relevant to insertion of INT such
   as trace type / instruction bitmap, max hop length, namespace ID, etc.
   This "TAM INT" object refers to an ACL or ACL group, which contains
   entries with a new ACL action attribute that is just a boolean,
   SAI_ACL_ENTRY_ATTR_ACTION_INSERT_INT. This action does not refer to
   the relevant attributes, instead one walks back to the "TAM INT" object
   that referenced this ACL. Since the reference is reversed, that
   proposal essentially uses a subset of ACL functionality instead of the
   full existing ACL functionality.

   With the other SAI TAM INT proposal, how are overlaps and conflicts
   addressed? Suppose the user wants to collect INT metadata in two
   different ways:

     a. For traffic with SAI_ACL_ENTRY_ATTR_FIELD_DST_IP matching
        10.20.0.0/16, use SAI_TAM_INT_ATTR_TRACE_VECTOR 0b1101 and
        SAI_TAM_INT_ATTR_MAX_HOP_COUNT 8.

     b. For all other traffic, with SAI_ACL_ENTRY_ATTR_FIELD_DST_IP
        matching 10.0.0.0/8, use SAI_TAM_INT_ATTR_TRACE_VECTOR 0b0011
        and SAI_TAM_INT_ATTR_MAX_HOP_COUNT 5.

   With the other SAI TAM INT proposal, this would require creation of
   two different objects sai_tam_int_obj1 and sai_tam_int_obj2.
   In sai_tam_int_obj1, SAI_TAM_INT_ATTR_ACL_OBJECT_LIST would refer to
   sai_acl_int_entry_obj1 which matches 10.20.0.0/16.
   In sai_tam_int_obj2, SAI_TAM_INT_ATTR_ACL_OBJECT_LIST would refer to
   sai_acl_int_entry_obj2 which matches 10.0.0.0/8.
   When a packet arrives with dest IP address matching 10.20.0.0/16, the
   priorities would have to be compared between two ACL entries in
   different SAI_TAM_INT_ATTR_ACL_OBJECT_LISTs in different
   sai_acl_int_entry_obj objects. The scope of comparison has now moved
   up to at least the TAM object level.

   With the approach proposed in this patch, only one ACL is needed,
   with at least two different ACL entries referring to two different
   "TAM INT Group" objects. Priorities are used to resolve overlaps and
   conflicts within the ACL.

   This approach removes several redundant objects and attributes such as
   the TAM INT flow object, counter object (because ACL rules can specify
   counters), and sample attribute (because ACL entries can directly refer
   to sample objects).

2. TAM INT domain boundaries and INT termination

   Typically INT type functionality is thought of as operating in a domain.
   When a packet first enters the domain, a decision is made whether to
   insert INT or just let the packet progress without INT. When a packet
   reaches the far edge, leaving the domain, then INT metadata is extracted
   to be processed or reported and the packet is progressed (if it is not
   a probe or a cloned packet) without any INT contents. In order to avoid
   leaking INT information outside of the domain, the edges of the domain
   must be properly configured.

   This proposal specifies a simple port list that indicates the edges of
   the domain. Whenever a packet is to be transmitted over one of these
   ports, INT must be removed before progressing the packet.

   The other SAI TAM INT proposal specifies INT termination actions by
   having the "TAM INT" object refer to an ACL including entries with a
   new boolean ACL action attribute SAI_ACL_ENTRY_ATTR_ACTION_DELETE_INT.
   There are two issues with such an approach:

     a. This introduces significant complexity by requiring an ACL to be
        specified, increasing the chances of operator errors leading to
        leaking of INT information outside of the domain.

     b. Typically in an ACL, only one action is applied, associated with
        the ACL entry with the highest priority. This suggests that
        separate ACLs should be defined, one with action INSERT_INT and
        another with action DELETE_INT. For example, if the parent TAM
        object is bound to a port, then traffic received on the port
        would be subject to INSERT_INT actions, while traffic transmitted
        out the port would be subject to DELETE_INT actions. This also
        calls into question whether the same "TAM INT" object should be
        defined for both insertion related attributes and deletion
        related attributes.

3. In this proposal, separate objects are proposed for separate
   collections of attributes with different granularity:

     a. "TAM INT Group" object is used to specify per flow insertion
        related attributes such as trace type / instruction bitmap,
        max hop length, namespace ID, etc. The "TAM INT Group" object
        is referenced from ACLs.

     b. A global "TAM INT" object is used to specify encapsulation
        aspects such as values that indicate presence of INT, which
        should be consistent across the entire INT domain. There is
        no good reason for these to vary across different subsets of
        traffic in a switch.

        The global "TAM INT" object is also used to specify the INT
        switch ID and the edge port list.

     c. Aspects related to report generation such as report format,
        collector addresses, etc should be grouped together as
        sink/terminator functionality that should not be coupled
        with INT insertion related attributes. See item 4 below.

   In the other SAI TAM INT proposal, all kinds of disparate attributes
   with different granularities are thrown into the same object
   definition. The resulting "TAM INT" object serves multiple purposes,
   including specification of trace vectors, flow lists, etc. Due to
   this, one TAM object referred to by a port may have many TAM INT
   objects for different flows, different trace vectors, etc. This will
   cause confusion when the same TAM INT object also specifies
   attributes that do not vary for different flows.

4. The TAM INT object also specifies the binding between TAM INT and
   the TAM collector and report objects.

   Upon termination of INT, it is not necessary for every packet to
   trigger a report. Receipt of such a packet should feed through event
   detection logic that determines whether to generate a report.

   This proposes to reuse the already existing TAM event object, treating
   receipt of packets with INT as events. The TAM event object already
   specifies reporting of events, referring to the TAM collector and
   report objects that apply whenever such an event occurs.

Gfrom2016 committed:
config lag neighbor and route according to config (#1553)

* config lag according to new config

Signed-off-by: zitingguo <zitingguo@microsoft.com>
Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* Add port1 route interface into default route config

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

harshitgulati18 committed:
Changes for supporting some PHY Diagnostics (#1527)

This change includes support for reading the following PHY layer diagnostics in SAI -

Per PMD Lane Rx Signal Detect
Per PMD Lane Rx Lock Status (aka CDR status)
PCS Rx Link Status
Per FEC Lane Alignment Marker Lock

chrispsommers committed:
Saithriftv2 fixes for extensions (#1533)

* [doc] Add SAI metadata introduction presentation (#1487)

Small presentation about SAI metadata features added in last time.

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Add GEN_SAIRPC_OPTS to pass flags to gensairpc.pl from top-level Makefile or meta/Makefile

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Add static_cast<> to sairpcgen templates to avoid type mismatches for SAI extensions.

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Add new env variable SAIRPC_EXTRA_LIBS which gets passed to final linkage of saiserver, so user can pass in additional libraries.

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Add experimental to include path.

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Add experimental headers to CPPFLAGS passed to ctypesgen.

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Add VLAN and PortChannel test plan (#1480)

* Add VLAN and PortChannel test plan

For vlan, it includes:
1. Vlan functionalities
   - Flooding
   - Forwarding
   - Trunk/Access
   - Tagging/Untagging(802.1Q)
   - VLAN interface (RIF/SVI)
2. SAI APIs operations
   - Vlan Counters
   - Vlan and member list operations

For PortChannel, it includes:
1. Lag functionalities
    - Loadbalance
2. Lag SAI APIs
    - create/check/remove lag and lag member

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* Update lag_test_plan.md

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* Update lag_test_plan.md

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* remove cases not related to T0

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* Update lag_test_plan.md

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* Add VLAN and Lag test plan and put topo config in seperated files

Signed-off-by: richardyu <richardyu@contoso.com>

* fix typo and more description more clear

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* fix lag test typo (#29)

Signed-off-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

* refactor vlan test plan and config

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* refactor lag test plan (#33)

Signed-off-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

* refactor pc and vlan test

Signed-off-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

* Improve pc and vlan test Doc (#35)

* fix typo

Signed-off-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

* remove some unnecessary comments

Signed-off-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

* reformat and adjust the description of the tests

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* set hash algorithm (#36)

Signed-off-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

* refine lag test (#37)

Signed-off-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

* correct vlan interface test (#38)

Signed-off-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

* workaround for the sequence number issue in github and correct typo
standardise the ip and mac naming in config

add lag test for ingress port indifferenate in lag hash
redesign the mac and ip namin convention and reformat
Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* adjust the case numbers and vlan id

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

Co-authored-by: zhoudongxu <1315364814@qq.com>
Co-authored-by: richardyu <richardyu@contoso.com>
Co-authored-by: zhoudongxu <Dongxu.Zhou@microsoft.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* [SAI-PTF] SAI PTF test structure for T0 test and VLAN,FDB sample cases (#1499)

* [SAI-PTF] Add SAI ptf test structure for T0 test

Create SAI PTF test structure for T0
1. Create configuration structure for different components configu
2. Add FDB, PORT, switch, vlan configurations
3. Create Test structure base on PTF
4. Add T0 test case samples for FDB

Test Done:
Checked the ports functionality with port flooding
Checked port forwarding base on FDB configurations

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* add vlan config and vlan test sample case

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* add dx010 config

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* [meta] Add sai_status_t to ancestry check (#1501)

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* reset sleep time during turning up ports (#1502)

Signed-off-by: Yang Wang <yangwang1@microsoft.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* [saineighbor] Bulk API support for Neighbor entries (#1504)

* Bulk support for Neighbor entries

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* add vlan testcases from Group6 to Group10 (#1506)

Signed-off-by: Yang Wang <yangwang1@microsoft.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Support counters on IP MC route entries. This is similar to the support added earlier for Route counters. (#1497)

Support counters on IP MC route entries. This is similar to the support added earlier for Route counters.

API sequence:

sai_attribute_t counter_attr;
attr.id = SAI_COUNTER_ATTR_TYPE;
attr.value.s32 = SAI_COUNTER_TYPE_REGULAR;
sai_object_id_t counter_id;
sai_status_t rc = sai_counter_api->create_counter(&counter_id, g_switch_id, 1, &attr);
sai_attribute_t ipmc_attr;
attr.id = SAI_IPMC_ENTRY_ATTR_COUNTER_ID;
attr.value.oid = counter_id;
rc = sai_ipmc_api->set_ipmc_entry_attribute(route_entry, ipmc_attr);
sai_stat_id_t stat_ids[] = { SAI_COUNTER_STAT_PACKETS, SAI_COUNTER_STAT_BYTES };
uint64_t stats[2];
rc = sai_counter_api->sai_get_counter_stats_ext(counter_id, 2, stat_ids, stats);

Signed-off-by: rck-innovium <rck@innovium.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Add thr missing API into saithriftv2 missing for syncd (#1484)

saithriftv1 offers start_sai_thrift_rpc_server API to start the saithrift server and syncd is currently using that API https://github.com/Azure/sonic-sairedis/blob/master/syncd/syncd_main.cpp#L60

But this API is missing from the for saithriftv2.

Made the relevant changes to provide a similar API for saithriftv2 library

The start_sai_thrift_rpc_server API currently resides in saiserver.cpp which is not included in librpcserver.a and is not available for the users of the static library. Thus moved the function definition into sai_rpc_frontend.cpp for including it in the static library.

Signed-off-by: Vivek Reddy Karri <vkarri@nvidia.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* add vlan test case (#1510)

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

add vlan test case

edit FilteringTest

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

edit vlan test case

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* add more waiting time for test (#1511)

* add more waiting time for test

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* Update fdb_configer.py

increase the waiting time for fdb entries update

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* remove useless change

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* add read config_db.json (#1519)

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

format

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

addomment for each function

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

add comment for each function

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* [Fix]Add necessary test data file for sai test (#1520)

In order to set up the port on different platforms, need to read the data for port configuration when turning on the admin state

This PR is a bug fix for adding the config_db into test resources folder.

Test Done:
Test within SONiC mgmt container

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Zitingguo/add lag test (#1522)

* add lag config

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* enable lag config in test base

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* fix lag config

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* add remove_lag_member in lag_configer

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* add sai_ipprefix() in utils

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* fix typo

Signed-off-by: zitingguo <736034564@qq.com>

* import LagConfiger in sai_test_base

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* add a simple lag test

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* fix typo

Signed-off-by: zitingguo <736034564@qq.com>

* fix a config error in lag

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* add lag lb test based on scr port

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* fix

Signed-off-by: zitingguo <736034564@qq.com>

* set lag v4 hash

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* add lag disable egress test case

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* add IndifferenceIngressPortTest

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* fix set hash function

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* separate route configuration from lag

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* disable setting lag hash

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* add virtual interface

Signed-off-by: Ziting Guo <zitingguo@microsoft.com>
Signed-off-by: zitingguo <736034564@qq.com>

* Config default v4&v6 route

Signed-off-by: zitingguo <736034564@qq.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* add teardown for testing multiple cases at a time (#1521)

* add read config_db.json

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

format

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

addomment for each function

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

add comment for each function

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* tear down

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Added missing dependencies required to run gensairpc.pl (#1492)

Fixes #1490

Co-authored-by: Chris Sommers <chrispsommers@gmail.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Update saitam.h (#1496)

Attribute is introduced to configure IPFIX template reporting interval as per the RFC7011.

https://datatracker.ietf.org/doc/html/rfc7011#section-8.4

Signed-off-by: Jai Kumar <jai.kumar@broadcom.com>
Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Pass SAIRPC_EXTRA_LIBS flag in Makefile

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

* Removed changes to include paths for SAI extension headers;
as it turns out these changes were not required as originally reported in https://github.com/opencomputeproject/SAI/issues/1531.

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

Co-authored-by: Kamil Cudnik <kcudnik@gmail.com>
Co-authored-by: Chris Sommers <chrispsommers@gmail.com>
Co-authored-by: Richard.Yu <richard.yu@microsoft.com>
Co-authored-by: zhoudongxu <1315364814@qq.com>
Co-authored-by: richardyu <richardyu@contoso.com>
Co-authored-by: zhoudongxu <Dongxu.Zhou@microsoft.com>
Co-authored-by: Yang Wang <yangwang1@microsoft.com>
Co-authored-by: Prince Sunny <prince.sunny@microsoft.com>
Co-authored-by: Ravindranath C K <rck@innovium.com>
Co-authored-by: Vivek R <vivekreddykarri98@gmail.com>
Co-authored-by: ms-junyi <106718431+ms-junyi@users.noreply.github.com>
Co-authored-by: zitingguo-ms <736034564@qq.com>
Co-authored-by: Jai Kumar <jai.kumar@broadcom.com>

richardyu-ms committed:
[SAI-PTF]Align the sai rpc front name with swss and syncd (#1552)

In Syncd starting process, it will try to start the switch and assign the gSwtichId to the sai implementation.
In the syncd rpc container, in order to get the swtich id after syncd initalization, we need to get and check the switch id.

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

zhoudongxu committed:
[SAI Test] Add L3 Neighbor Test (#1538)

* add l3 neighbor test

Signed-off-by: zhoudongxu <Dongxu.Zhou@micorsoft.com>

* impprove nexhop diff prefix  test

Signed-off-by: zhoudongxu <Dongxu.Zhou@micorsoft.com>

* add nohost route comments

Signed-off-by: zhoudongxu <Dongxu.Zhou@micorsoft.com>

Co-authored-by: zhoudongxu <Dongxu.Zhou@micorsoft.com>

zhoudongxu committed:
Add L3 router interface test (#1554)

* add l3 router interface test

Signed-off-by: zhoudongxu <Dongxu.Zhou@micorsoft.com>

* add blank

Signed-off-by: zhoudongxu <Dongxu.Zhou@micorsoft.com>

* add new line

Signed-off-by: zhoudongxu <Dongxu.Zhou@micorsoft.com>

Co-authored-by: zhoudongxu <Dongxu.Zhou@micorsoft.com>

chrispsommers committed:
Check for NULL APIs returned by sai_api_query() before calling. (#1558)

Signed-off-by: Chris Sommers <chrispsommers@gmail.com>

Co-authored-by: Chris Sommers <chrispsommers@gmail.com>

richardyu-ms committed:
[SAI-PTF]Add more description in SAI test config (#1560)

* [SAI-PTF]Fix SAI test config

Fix SAI test config, fix the next hop ips, they should be match the IP address in neighbor
Add more mac address for port 9 ~ 16

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* Lag test plan: Add ip v6 branch fix a description in test case

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* fix defect for config with VLAN SVI port and RIF

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* Add one more group of data for  neighbor

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* fix a typo

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* change IP for lags

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* remove unnecessary IP address

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

zhoudongxu committed:
Disable cpu queue stat  check (#1561)

* Disable cpu queue stat  check

Signed-off-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

* add bug comment

Signed-off-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

marian-pritsak committed:
Tunnel UDP SRC PORT security (#1455)

 Drop tunnel packets with not allowed UDP source port
     *
     * Upon enabling this feature, if the tunnel packet ingresses with
     * UDP source port outside of range defined for this tunnel, it
     * will be dropped.

Signed-off-by: Marian Pritsak <marianp@mellanox.com>

allen-xf committed:
[SAI-PTF] Refactor code for more generic service and dut data modeling   (#1556)

* rebase master and edit config

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* refactor lag cases

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* fix bug and autopip

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* format

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* edit vlan case followed by dut and device

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* refactor_rif_ipv4

* edit remaining cases and add typing

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* add rif_ipv6_test

* refactor l3 neighbor

* (fix) modify port1_rif to port0_rif

* (fix) port_list[0]

* fdb case

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* test the first half of the fdb case

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* fdb refactor

* (fix) self.servers[1][index-1].mac

* (fix) lag1_neighbor

* fix sanity test

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>
Co-authored-by: zhiyuan-ms <liangzy512@gmail.com>
Co-authored-by: Richard.Yu <richard.yu@microsoft.com>

richardyu-ms committed:
[SAI-PTF] Refactor test structure, add data module and add more method (#1564)

* [SAI-PTF] Refactor test structure, add data module and add more method

Description of the PR
Refactor test structure, add data module and add more method

Detail
Add data module in the test structure
Create data module when init common config
Add more method in L3 for testing
Add more comment
Refactor code

Test done:
Local test, run rif, neighbor and lag tests

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* change device list to device dict, add lag obj to device, add more comment and refactor code

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* sepeate data model, add more data module ecmp and nexthop, add more method for l3 cases

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* add more L3 method

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* add L3 method for different situation, port, bridge port, vlan and acl

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* commit for testing

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* fix defect and reformat code

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* correct code base on the code review

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

richardyu-ms committed:
[SAI-PTF] FIX - remove and cleanup is_ipv4 (#1566)

Remove and clean up is_ipv4.
This parameter is not used any more after refactor, and this parameters cause running error.

Detail:
Remove all the place referencing this parameter

Test done:
Test with FDB case

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

Gfrom2016 committed:
fix bug when dst prefix is 32 or 128 for v6 (#1567)

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

Gfrom2016 committed:
Change mac address representation and fix bugs (#1568)

* represent mac address in hex

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* fix assign neighbor list bug

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* Set nbr_entry_v6 None when ipv6 address no exist

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

JaiOCP committed:
Distributed HW resource management infrastructure (#1528)

This PR introduces an infrastructure to be able to query the HW resource for its ingress, egress or both stages i.e. that if physical resource belongs to ingress pipeline, egress pipeline or is a shared resource across ingress/egress stages of pipeline.

Additionally hostif trap group and policer pool object is enhancement with an attribute to specific which stage of the pipeline trap group and/or policer pool belongs.

For details and sample workflow please refer to the spec.

Signed-off-by: Jai Kumar <jai.kumar@broadcom.com>

Gfrom2016 committed:
reconfig servers list start from 0 instead of 1 (#1572)

* reconfig servers list start from 0 instead of 1

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* remove unused variable in Device()

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* create 101 server and t1 device, change lag neighbor ip according to
config

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* Separate create default loopback from default virtual router

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* fix neighbor test cases

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* change fun name and add help doc

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* Add help doc for t0_route_config_helper

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>
Co-authored-by: zhoudongxu <Dongxu.Zhou@microsoft.com>

Gfrom2016 committed:
use skip mark to skip unsupported test case (#1576)

Signed-off-by: zitingguo <zitingguo@microsoft.com>
Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

Signed-off-by: zitingguo <zitingguo@microsoft.com>
Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

allen-xf committed:
[SAI-PTF]persist and read dut (#1569)

* persist and read dut

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* fix persist

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* vlan user super teardown

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* persist server and t1

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

richardyu-ms committed:
[SAI-PTF]Data module for warm reboot persist and load the data (#1577)

* [SAI-PTF]Data module for warm reboot persist and load the data

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* refactor the port object

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* replace all the port related data to port object

fix bugs
cleanup code
fix bug

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* use predefined config data in  test cases

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* add config and correct test cases
rollback the class renaming
refactor code, simplify the structure

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* fix cases issue use the dynamic dev port index but not the static number
change variable name base on comment in PR
change dev port index

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

Gfrom2016 committed:
[SAI-PTF]Refactor lag cases (#1578)

* refactor lag cases

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* restore lag member egress attr

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

* change all port index to dev_port_index

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

Signed-off-by: zitingguo-ms <zitingguo@microsoft.com>

allen-xf committed:
Refactor vlan cases (#1579)

* refactor all vlan cases:change all port index to dev_port_index

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

* fix bug

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

Signed-off-by: ms-junyi <t-junyixiao@microsoft.com>

richardyu-ms committed:
Add l3 test plan (#1505)

* add neighbor, nexthop group, route test cases change config

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* add more cases and change mode case depend on existing config
add more constraint and description for test cases

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

richardyu-ms committed:
[SAI-PTF] Create Port, bridge port and host interface base on the port_config.ini (#1584)

* [SAI-PTF] Create Port, bridge port and host interface base on the port_config.ini

In order to create the port, bridge port and host interface on different kinds of the configurations of the device, depends on port_config.ini to create them.

This change include the modifications
1. read and parse the port_config.ini file
2. sorting the port list base on the config
3. create bridge and host interface base on the config
4. add a decorator for the data object to auto generate the __str__ for debugging

Test done:
test cases on duts

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

* fix a bug, assign the bridge port id

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

kannankvs committed:
release notes for SAI 1.10.2 version (#1545)

This PR is reference to PR1523 as the previous PR has a DCO error. Raised this new PR. The previous PR has been approved by the community member for the SAI release notes for 1.10.2 version

Signed-off-by: kannankvs <kannan_kvs@dell.com>

richardyu-ms committed:
[SAI-PTF]Skip Test cases ny saise SKipTest (#1593)

Prepare the structure to skip test cases base on different ASIC.
In this modification, just put the skip logic to the super class, will
add more code to skip cases base on different platform and ASIC
infomation.

Test done:
Test in DUT.

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>
